hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/security/token/delegation/AbstractDelegationTokenSecretManager.java

```diff
@@ -726,4 +732,41 @@ public TokenIdent decodeTokenIdentifier(Token<TokenIdent> token) throws IOExcept
     return token.decodeIdentifier();
   }
 
+  /**
+   * Return top token real owners list as well as the tokens count.
+   *
+   * @param n top number of users
+   * @return map of owners to counts
+   */
+  public List<NameValuePair> getTopTokenRealOwners(int n) {
```

Comment:
will this get pretty expensive if there are lots of tokens stored? as every metrics pull needs to iterate through all tokens.

Reply:
I had similar thoughts as well but didn't come up with a better way. In namenode TopN metrics, it is doing so as well just at a lower frequency like every 5/15/25 minutes. We can potentially do that by reducing the metric reporting frequency.
I also checked modern CPU for looping over 1M, which takes about 1ms-10ms.
Another one would be to maintain a data structure to dynamically maintain the ordering of users and edit ordering per getdelegationtoken and per canceldelegationtoken like stream processing. I am not sure about the cost overall since in reality we generally have < 1 Million tokens.

Reply:
Can we update the `TopN` queue when creating/deleting tokens? we are just paying an extra constant cost for updating that which I think is fine. Even though it is using concurrent hashmap, I'm not sure how much performance impact will be if one thread is iterating over the key set while others want to updating the map.

Reply:
I am fine with it. We also need to add an initialization step to make sure this structure has the initial information from currentTokens.